### PR TITLE
Integrate Socket IO with Frontend

### PR DESCRIPTION
**Pull request by bharathcs**
# Add Jotai atoms and Socket.IO integration around Innkeeper

Elroy & Kevin wanted to use Jotai, so I looked up that stuff. Since a lot of the performance benefits come from
using Jotai utilities to neatly separate read / write dependencies down to the nesting / filtering, it would make
the most sense to abstract away the state management (including Socket IO work) away from the room page.

What works rn: Use the room page to connect, make matching request, and update the text editor. Run postman and act as the collaborator to make a matching request to complete the pair, and to watch the updates sent out by the browser. Remember to use something other than "user_a" in Postman.

---

### Atoms

Jotai manages state and automatically tracks read / write dependencies to avoid too many renders.

Basic Jotai atoms have been constructed, with the largest of which being roomState.
Dependent Jotai atoms have been constructed to filter / nest / focus on the correct values inside roomState.

### Socket IO

Handler is set up using a hook so that the Jotai-managed socket object will be initialised when the component loads,
and disconnected and set to null when the component unloads. (i.e. when /room page loads and is closed)

In the hook, handlers are attached to listen to server events and update Jotai-managed state (Jotai atoms of roomState, isConnected etc).
Additional Jotai atom was added as a utility function to support emitting events as simply as possible from the frontend

### Room Page

To work with how dumb the innkeeper auth logic is, a hardcoded 'user_a' is sent as auth token, and this is used to set up socket.

- If not connected, shows a loading page.
- If not matched, shows a placeholder lobby question (easy, medium, hard)
- Once matched, shows the question pane and code editor pane.

Use console and a Postman to verify that all the messages are being sent as expected.

---

- Add socket io deps
- Add socket io lib and share types
- Update CodeEditor for basic use
- Add jotai
- Fix build error
- Update innkeeper api type
- Fix bug from combining namespaces: use lobby room
- Add useInnkeeper hook and room-related Jotai atoms

### CODE REVIEW

File: frontend/package.json
Comment (bharathcs) [inline frontend/package.json:1]: Had to add these additional npm pkgs: `bufferutil, optics-ts, utf-8-validate, ws` for some reason. I kept having build errors until all were added.

Comment (sivayogasubramanian) [reply frontend/package.json:1]: You need to add the new env variable to the CI files as well as the github secrets

File: frontend/.env.local.example [hunk 1]
```diff
@@ -6,4 +6,5 @@ NEXT_PUBLIC_FIREBASE_PROJECT_ID=
 NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET=
 NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID=
 NEXT_PUBLIC_FIREBASE_APP_ID=
-NEXT_PUBLIC_GOOGLE_CLIENT_ID=
\ No newline at end of file
+NEXT_PUBLIC_GOOGLE_CLIENT_ID=
+NEXT_PUBLIC_PEERPREP_INNKEEPER_SOCKET_URL=
```
Comment (sivayogasubramanian) [inline frontend/.env.local.example:11]: You need to add the new env variable to the CI files as well as the github secrets

Comment (bharathcs) [reply frontend/.env.local.example:11]: Thanks for the catch: [Update env for github workflows](https://github.com/PeerPrep/peerprep/pull/53/commits/39da179f508d34ad6befceada5f9d8f323076885)

**Review by sivayogasubramanian — APPROVED** [review]

**Pull request merged**